### ISSUE TITLE
BUG: Dask's _parse_gufunc_signature can't parse a singature returned by numpy 1.16rc1

### ISSUE THREAD

**Issue by detrout** [numpy/numpy#12712]
Hello,

The Debian numpy maintainer filed a a bug against my packaging of Dask (<a href="https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=918204">Debian Bug 918204</a>) because some of dasks's tests failed. One of the failures seems to be related to the changes for https://www.numpy.org/neps/nep-0020-gufunc-signature-enhancement.html

(relevant traceback from the dask tests: <a href="https://ci.debian.net/data/autopkgtest/unstable/amd64/d/dask/1667144/log.gz ">here</a> (search for "test_matmul ")
```_________________________________ test_matmul __________________________________

    @pytest.mark.skipif(sys.version_info < (3, 5),
                        reason="Matrix multiplication operator only after Py3.5")
    def test_matmul():
        x = np.random.random((5, 5))
        y = np.random.random((5, 2))
        a = from_array(x, chunks=(1, 5))
        b = from_array(y, chunks=(5, 1))
        assert_eq(operator.matmul(a, b), a.dot(b))
        assert_eq(operator.matmul(a, b), operator.matmul(x, y))
>       assert_eq(operator.matmul(a, y), operator.matmul(x, b))

/usr/lib/python3/dist-packages/dask/array/tests/test_array_core.py:849: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3/dist-packages/dask/array/core.py:1054: in __array_ufunc__
    **kwargs)
/usr/lib/python3/dist-packages/dask/array/gufunc.py:264: in apply_gufunc
    input_coredimss, output_coredimss = _parse_gufunc_signature(signature)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

signature = '(n?,k),(k,m?)->(n?,m?)'
```

The signature  '(n?,k),(k,m?)->(n?,m?)' looks like one of the examples from NEP 20, but I tried looking at Numpy's _parse_gufunc_signature to see how to update Dask's version and discovered that numpy's doesn't seem to parse  '(n?,k),(k,m?)->(n?,m?)' either.

I looked for a unit test and couldn't find the above signature test, so I added it to commit `30eae3dd24402ca07cd56cae6db646ad61e5d7f1` (not in NumPy: mattip)
```
--- a/numpy/lib/tests/test_function_base.py
+++ b/numpy/lib/tests/test_function_base.py
@@ -1362,6 +1362,8 @@ class TestVectorize(object):
                      ([('x',)], [('y',), ()]))
         assert_equal(nfb._parse_gufunc_signature('(),(a,b,c),(d)->(d,e)'),
                      ([(), ('a', 'b', 'c'), ('d',)], [('d', 'e')]))
+        assert_equal(nfb._parse_gufunc_signature('(m?,n),(n,p?)->(m?,p?)'),
+                     ([('m?', 'n'),('n','p?')],[('m?', 'p?')]))
         with assert_raises(ValueError):
             nfb._parse_gufunc_signature('(x)(y)->()')
         with assert_raises(ValueError):
```

and `python3 runtests.py` fails.

```
___________________ TestVectorize.test_parse_gufunc_signature ___________________

self = <numpy.lib.tests.test_function_base.TestVectorize object at 0x7f8de58c1f60>

    def test_parse_gufunc_signature(self):
        assert_equal(nfb._parse_gufunc_signature('(x)->()'), ([('x',)], [()]))
        assert_equal(nfb._parse_gufunc_signature('(x,y)->()'),
                     ([('x', 'y')], [()]))
        assert_equal(nfb._parse_gufunc_signature('(x),(y)->()'),
                     ([('x',), ('y',)], [()]))
        assert_equal(nfb._parse_gufunc_signature('(x)->(y)'),
                     ([('x',)], [('y',)]))
        assert_equal(nfb._parse_gufunc_signature('(x)->(y),()'),
                     ([('x',)], [('y',), ()]))
        assert_equal(nfb._parse_gufunc_signature('(),(a,b,c),(d)->(d,e)'),
                     ([(), ('a', 'b', 'c'), ('d',)], [('d', 'e')]))
>       assert_equal(nfb._parse_gufunc_signature('(m?,n),(n,p?)->(m?,p?)'),
                     ([('m?', 'n'),('n','p?')],[('m?', 'p?')]))

self       = <numpy.lib.tests.test_function_base.TestVectorize object at 0x7f8de58c1f60>

numpy/lib/tests/test_function_base.py:1365: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

signature = '(m?,n),(n,p?)->(m?,p?)'

    def _parse_gufunc_signature(signature):
        """
        Parse string signatures for a generalized universal function.
    
        Arguments
        ---------
        signature : string
            Generalized universal function signature, e.g., ``(m,n),(n,p)->(m,p)``
            for ``np.matmul``.
    
        Returns
        -------
        Tuple of input and output core dimensions parsed from the signature, each
        of the form List[Tuple[str, ...]].
        """
        if not re.match(_SIGNATURE, signature):
            raise ValueError(
>               'not a valid gufunc signature: {}'.format(signature))
E           ValueError: not a valid gufunc signature: (m?,n),(n,p?)->(m?,p?)

signature  = '(m?,n),(n,p?)->(m?,p?)'

numpy/lib/function_base.py:1794: ValueError
```

Should `_parse_gufunc_signature` handle the signature with question marks? If yes, should there be a test for that case?

Edit: noted that commit is not in NumPy (mattip)

**Comment by tylerjereddy** [comment]
@shoyer Seems to have indicated in that cross-linked issue that this is a Dask matter, but I'll let someone else chime in too since I know i.e. @mattip has done lots of work on this.

**Comment by mattip** [comment]
Dask must update its parsing. This is not a NumPy issue, but it is good to know that Dask does not yet handle NumPy's 1.16 change of `matmul` into a ufunc.

**Comment by shoyer** [comment]
Well, the short answer is that I wrote the gufunc signature parser for `np.vectorize` before we came up with optional dimensions, and then we copied it into dask.

`np.vectorize` still doesn't support optional dimensions, so it wouldn't really make sense to support it in this parser (which is specifically for `np.vectorize`). `_parse_gufunc_signature` itself isn't a public API. That said, arguably we should be raising a better error message, something like "Optional dimensions are not yet supported by `np.vectorize`."

This does point a hazard of creating new gufuncs: old code can break, because `np.matmul()` now calls `__array_ufunc__` overrides. Hopefully we'll have a more graceful transition in the future now that we'll have `__array_function__`, but in this particular edge case I don't think there's much we could have done better.

**Comment by mattip** [comment]
> copied it into dask

Rather than parsing, couldn't Dask look at the created and publicly available `matmul` `UFuncObject`? That said, it would be better to extend the `UFuncObject` with some python-level field access wrappers since the object itself can mutate.

**Comment by shoyer** [comment]
Let me try to unpack what happened here.

1. Dask supports `@` via `__matmul__` and `__rmatmul__`. Dask did not support `np.matmul` in its `__array_ufunc__` method (because this wasn't a thing).
2. `numpy_array @ dask_array` used to work via converting `dask_array` to a numpy array with `__array__()`; note that dask operations should be lazy and take precedence over numpy operations, but this didn't used to be possible (due to https://github.com/numpy/numpy/issues/9028). But apparently Dask's test suite ended up inadvertently relying on this (broken) behavior.
3. Now, calling `numpy_array @ dask_array` calls `__array_ufunc__` internally, because `np.matmul` is a ufunc and numpy uses `__array_ufunc__` instead of calling `__rmatmul__`. This now results in an error message, due to (1).

I think we're in the clear from the NumPy side -- it's not really preferrable to go back to the previous semi-broken behavior.

**Comment by detrout** [comment]
I agree Dask needs to be updated. (Thank you @shoyer ) but shouldn't numpy's version of parse_gufunc_signature also handle '(m?,n),(n,p?)->(m?,p?)'.

numpy.lib.function_base._parse_gufunc_signature('(m?,n),(n,p?)->(m?,p?)') raises a ValueError.

To figure out what to do for dask, I tried looking for gufunc related tests in numpy that handled the optional arguments, and didn't find any.

**Comment by mattip** [comment]
Ahh, indeed that is a bug. Why do we have two implementations of signature parsing? For tests, see `test_ufuncs.py` [test_signature5](https://github.com/numpy/numpy/blob/v1.16.0rc2/numpy/core/tests/test_ufunc.py#L343) and `test_signature8] which use the C parser.

**Comment by mhvk** [comment]
@mattip - See https://github.com/numpy/numpy/issues/12712#issuecomment-452981768 for why this other parser is there; this is definitely not a public method, so we're not beholden to anything - though having something exposed is certainly not crazy! In any case, even just from an internal numpy perspective it would be good to use the C parser internally for that (and let `vectorize` error on flexible dimensions). But probably best to make this a new issue!

**Comment by detrout** [comment]
should _parse_gufunc_signature's docstring mention that there's another gufunc parser or that this private function is only intended for use by np.vectorize?

**Comment by mattip** [comment]
Although we are not required to, it means `np.vectorize` only can parse a portion of the full numpy signatures. Extending the regexp is trivial, `_DIMENSION_NAME = r'\w+'` should become `_DIMENSION_NAME = r'\w+\?{,1}'`, but then I am not sure what the implications are for actual gufuncs defined this way.

**Comment by Zac-HD** [comment]
Hi all!  I'm working on a Hypothesis strategy to generate input shapes for ufuncs - https://github.com/HypothesisWorks/hypothesis/issues/2174 - which means I need to parse NEP20 signatures.

- Is the best way forward just to write our own regex-based parser?  If not, what should we use instead?  I don't mind if this feature explicitly requires e.g. Numpy >= 1.17 :smile: 
- Is it valid under NEP 20 for a dimension to have both broadcastable (`|1`) and optional (`?`) suffixes?

**Comment by shoyer** [comment]
Per the note at the top of the NEP, we didn't end up accepting
"broadcastable" dimensions.

**Comment by mattip** [comment]
We have not written a python-level parser, nor have we exposed the c-based parser. There was a request for this as part of a python-level wrapper for gufuncs in gh-14020.

**Comment by Zac-HD** [comment]
No worries, it sounds like copy-pasting amd extending `_parse_gufunc_signature` is the way to go then and I'm comfortable doing that. Thanks for the quick feedback everyone 🙂
